### PR TITLE
Feat/better storage

### PR DESCRIPTION
**Pull request by deadbelly**
### What Type of PR Is This?
- [x] New Feature 
- [x] Bug Fix
- [ ] Testing
- [x] Refactoring/Clean Up 

#### What does this branch do? 
This branch handles local storage for offline use more cleanly by utilizing a custom react hook. This hook is called `useDataStorage`, and it calls two `useEffect`s. The first sets a component’s state to specific stored data if the client is offline and if the data is available. The second returns a callback function that runs if the component will unmount. This callback saves the components state to local storage with unique keys.

While working on this branch I discovered a pretty significant bug: when the user creates a tour while offline, that tour becomes permanently bugged because it never creates a `Plan` table in the database. For now I’ve disabled the create tour button while offline, but I think there may be a better solution in the future. 

I’m curious about using custom hooks for fetching data when a component mounts. The patterns for each component are similar enough that I think one hook might be able to handle it but there are a few edge cases that may be difficult to accommodate. 
#### Have you tested your branch? How?
User testing on mobile and desktop.
#### Do all existing unit and integration tests pass? 
N/A
#### Does this branch require special installation steps? 
Please run `npm i`, then delete your browser's cache and local storage for the site, then uninstall your service worker.

### CODE REVIEW

File: src/customHooks/useDataStorage.js
Comment (tashiad) [inline src/customHooks/useDataStorage.js:19]: Looks like `'i' is declared but its value is never read` here

File: src/components/Form/TourForm.tsx
Comment (tashiad) [inline src/components/Form/TourForm.tsx:112]: Should delete this `console.log`

File: src/customHooks/useDataStorage.js
Comment (deadbelly) [reply src/customHooks/useDataStorage.js:19]: You're right that's a relic from an earlier version. Deleted.

File: src/components/Form/TourForm.tsx
Comment (deadbelly) [reply src/components/Form/TourForm.tsx:112]: Gone.

**Review by tashiad — CHANGES_REQUESTED** [review]
These are some really cool changes, Boone! I didn't know custom hooks were a thing.

I caught a couple of things while testing. Once these have been fixed I can merge.

- [ ] "Created by" isn't showing up on current tour cards like it is on past tour cards. I noticed that this has been commented out on lines 29-30 of `CurrentTourCard`
- [ ] If I try to fill out a form field in offline mode it doesn't save. This is after following the installation/clearing cache/service worker instructions in your PR. See screenshot of errors (the first field was filled out online, the second field should have had TESTTEST show up after clicking save but it doesn't):
Screen Shot 2021-05-18 at 1 18 09 PM

**Comment by deadbelly** [comment]
I'll look at the first bug right away, I'm not 100% sure why that line is commented out.

As for the second bug I can't seem to replicate it. I just want to double check: did you also delete local storage in addition to the cache and sw? I don't think that's the issue but wanted to check. 

From your console it looks to me like the service worker isn't running, period. At some point last week Rachel and I noticed that when the service worker updates you sometimes had to both clear the worker and reload, and then reload again for it to run on browsers that already had it installed. Is it possible that another reload fixes the issue? 

If that is the problem we should take a look at that in a separate branch.

**Comment by tashiad** [comment]
**QA Steps:**
1. Clear Local Storage & Session Storage
2. Clear Cache Storage
3. Unregister Service Worker
4. Reload
5. Reload again
6. Go to a current tour and save an edit to a text field while online.
7. Go offline, save an edit to another text field.
8. Go to past tours and then back to the tour I made the edit on. Edit has been successfully saved.
9. Go back online. Check to see if offline edit persists. It's there!

I think you're right. The above steps were successful. But when I went through the process again and skipped step 5, step 9 failed.

If we think this will work well on the deployed site, I'm ok with merging! But ideally would like to get that "created by" code back on the current tour cards.

**Comment by deadbelly** [comment]
Got the `created by` block working again. Sorry it took me so long- I had some issues unregistering the service worker for some reason. I think this branch is good to merge whenever.

**Pull request merged**
